### PR TITLE
Update dispatch filter for recurrence and end time [No Enum Alternative]

### LINKED ISSUE #161: Extend the dispatch API to allow filtering of reoccurring and non-reoccurring dispatches by their end time
### What's needed?

Can the filtering be extended to allow filtering by end time and duration in order to determine if a dispatch has finished or not?

### Proposed solution

```Diff
index 4166925..6546d5f 100644
--- a/proto/frequenz/api/dispatch/v1/dispatch.proto
+++ b/proto/frequenz/api/dispatch/v1/dispatch.proto
@@ -149,6 +149,18 @@ message DispatchDetail {
   google.protobuf.Timestamp modification_time = 4;
 }
+ // Enum for the type of dispatch based on its recurrence.
+ enum RecurrenceType {
+   // UNSPECIFIED: Recurrence type is not specified.
+   RECURRENCE_TYPE_UNSPECIFIED = 0;
  
+   // NON_RECURRING: Dispatches that do not recur.
+   RECURRENCE_TYPE_NON_RECURRING = 1;
  
+   // RECURRING: Dispatches that recur.
+   RECURRENCE_TYPE_RECURRING = 2;               
+ }

// Filter parameter for specifying multiple time intervals
message TimeIntervalFilter {
-  // Filter by start_time >= this timestamp
+  // Filter by time >= this timestamp.
-  google.protobuf.Timestamp start_from = 1;
+  google.protobuf.Timestamp from = 1;

-  // Filter by start_time < this timestamp
+  // Filter by time < this timestamp.
-  google.protobuf.Timestamp start_to = 2;
+  google.protobuf.Timestamp to = 2;

-  // Filter by recurrence.end_criteria.until >= this timestamp
-  google.protobuf.Timestamp end_from = 3;

-  // Filter by recurrence.end_criteria.until < this timestamp
-  google.protobuf.Timestamp end_to = 4;
}

// Parameters for filtering the dispatch list
message DispatchFilter {
-  // Filter by component ID or category
+  // Optional filter by component ID or category.
  repeated ComponentSelector selectors = 1;

-  // Filter by time interval
-  // If no interval is provided, all dispatches starting from the
-  // current timestamp will be included.
-  TimeIntervalFilter time_interval = 2;

-  // Filter by active status
+  // Optional filter by active status.
  // If this field is not set, dispatches of any active status will be included.
-  optional bool is_active = 3;
+  optional bool is_active = 2;

-  // Filter by dry run status
+  // Optional filter by dry run status.
  // If this field is not set, dispatches of any dry run status will be included.
-  optional bool is_dry_run = 4;
+  optional bool is_dry_run = 3;

+  // Optional filter by reoccurring and none-reoccurring dispatches.
+ // If not specified both types will be returned.
+  optional RecurrenceType recurrence_type = 4;

+  // Optional filter by start time.
+  // If no interval is provided, all dispatches will be returned.
+  TimeIntervalFilter start_time_interval = 5;

+  // Optional filter by end time
+  // Filter dispatches based on their explicit or estimated end time.
+  TimeIntervalFilter end_time_interval = 6;

+  // Optional filter by update time
+  TimeIntervalFilter update_time_interval = 7;
}
```

### Use cases

_No response_

### Alternatives and workarounds

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Marenz**
Alternative to #183

fixes #161

[Sanctioned by Tom :tm: ](https://github.com/frequenz-floss/frequenz-api-dispatch/issues/161#issuecomment-2214376199)

### CODE REVIEW

**Comment by Marenz** [comment]
Gimme, gimme, gimme a review after midnight,
Won't somebody help me check these lines of code?
Gimme, gimme, gimme a review after midnight,
Take a look and help me lighten up my load.

Opened up the pull request,
Commits are tidy, code at its best.
All the CI checks have passed,
Need approval, and need it fast.

Staring at the screen so long,
Hoping you'll come along.

File: proto/frequenz/api/dispatch/v1/dispatch.proto [hunk 1]
```diff
@@ -151,17 +151,68 @@ message DispatchDetail {
...
-  // Filter by start_time >= this timestamp
-  google.protobuf.Timestamp start_from = 1;
+  // Filter by time >= this timestamp.
+  google.protobuf.Timestamp from = 1;
 
-  // Filter by start_time < this timestamp
-  google.protobuf.Timestamp start_to = 2;
+  // Filter by time < this timestamp.
+  google.protobuf.Timestamp to = 2;
+}
```
Comment (ela-kotulska-frequenz) [inline proto/frequenz/api/dispatch/v1/dispatch.proto:159]: estimated_end_time will be calculated for all dispatches - not only reoccurring
So it make sense to leave the end_to, end_from here. Please look at:
https://github.com/frequenz-floss/frequenz-api-dispatch/pull/186/commits/92f675dd24686ee3b874862985992fff00043c06

Comment (Marenz) [reply proto/frequenz/api/dispatch/v1/dispatch.proto:159]: It's not removed, just spilt up into start interval and end interval, see: https://github.com/frequenz-floss/frequenz-api-dispatch/pull/184/files#diff-b2f8ee9bf7dcd2f8c9a8d9434771f0229aca74577fdda0831e26be679fae6cdbR212

Comment (ela-kotulska-frequenz) [reply proto/frequenz/api/dispatch/v1/dispatch.proto:159]: Oww that is right, sorry for misleading comment . 
For some reason I though it is now part of RecurrenceFilter, but it is not ...
Anyway approve :)

**Review by ela-kotulska-frequenz — APPROVED** [review]

**Pull request merged**